snapshot/snapshot.sh

```diff
@@ -0,0 +1,328 @@
+
+    # Add "runtimeClassName: kontain" to the pod or deployment.
+    $YQ '(.spec.runtimeClassName = "kontain")' -i $TARGET_POD_YAML
+ 
+    kubectl apply -f $TARGET_POD_YAML >& /dev/null
+
+    echo "done"
+}
+
+function prepare_snapshot_pod() {
+
+    echo -n "Preparing snapshot maker....."
+    cp $SNAPSHOT_YAMLFILE $MAKER_POD_YAML
+
+    #make sure pod name is what this cript expects 
+    pod=$MAKER_POD_NAME $YQ  -i '(.metadata.name=strenv(pod))' $MAKER_POD_YAML
+
+    # pass port and ip to the control pod     
+    kubectl wait --for=condition=Ready pod -n $KONTAIN_NAMESPACE $TARGET_POD_NAME >& /dev/null
+    status=$(kubectl get pod  -n $KONTAIN_NAMESPACE $TARGET_POD_NAME -o jsonpath="{.status.phase}")
+    if [ "$status" != "Running" ]; then 
+        echo "Error starting target pod"
+        exit 1
+    fi
+
+    ip=$(kubectl get pod -n $KONTAIN_NAMESPACE  $TARGET_POD_NAME -o jsonpath='{.status.podIP}')
+    port=$(kubectl get pod -n $KONTAIN_NAMESPACE $TARGET_POD_NAME  -o jsonpath='{.spec.containers[0].ports[0].containerPort}')
+    port=${port:=80}
+
+    declare -A  extra_kontain_envs
+    extra_kontain_envs[IP]=$ip
+    extra_kontain_envs[PORT]=$port
+
+    add_envs $MAKER_POD_YAML extra_kontain_envs
+
+    # assign pod to specific node
+    set_node_selector $MAKER_POD_YAML
+
+    kubectl apply -f snapshot_entrypoint.yaml >& /dev/null
+    kubectl apply -f $MAKER_POD_YAML >& /dev/null
+
+    # wait for manager pod  to be ready, i.e untill snapshot is made 
+    kubectl wait --for=condition=Ready pod -n $KONTAIN_NAMESPACE $MAKER_POD_NAME  >& /dev/null
+    status=$(kubectl get pod  -n $KONTAIN_NAMESPACE $MAKER_POD_NAME -o jsonpath="{.status.phase}")
+    if [ "$status" != "Running" ]; then 
+        echo "Error starting maker pod"
+        exit 1
+    fi
+
+    echo "done"
+}
+
+function make_image() {
+
+    echo -n "Generating a new snapshot-based image....."
+
+    mkdir -p $IMAGE_DIR
+    kubectl cp $KONTAIN_NAMESPACE/$MAKER_POD_NAME:$SNAPDIR_CONTAINER $IMAGE_DIR >& /dev/null
+
+    # Create a new container image to run the snapshot
+    # This needs to handle differing snapshot filenanes kmsnap vs kmsnap.XXX.NNN
+    # Build dockerfile
+    # Build container image
+    # Push the container to a remote repository
+    SNAPFILE=$(basename $IMAGE_DIR/kmsnap*)
+
+    cat <<EOF >$SNAP_DOCKER_FILE
+FROM $IMAGE_NAME
+COPY $SNAPFILE /kmsnap
+CMD ["/kmsnap"]
+EOF
+
+    SNAP_IMAGE_NAME=$SNAPREPO/$(basename $IMAGE_NAME)_snap:latest
+
+    chown $(id -u):$(id -g) $IMAGE_DIR/$SNAPFILE
+    chmod 755 $IMAGE_DIR/$SNAPFILE
+
+    docker build -t $SNAP_IMAGE_NAME -f kmsnap.dockerfile $IMAGE_DIR >& /dev/null
+
+    echo "done"
+}
+
+function upload_image() {
+    echo -n "Uploading image to dockerhub....."
+
+    if [ -n "$docker_uname" ]; then
+        cat $docker_passfile | docker login --username $docker_uname --password-stdin
+    fi;
+
+    docker push $SNAP_IMAGE_NAME >& /dev/null
+
+    echo "Image $SNAP_IMAGE_NAME has been created"
```

Comment:
Replace "created" with "uploaded" ?

Reply:
done. Will also print whatever error docker  gives us in case of error